src/Microsoft.Health.Fhir.TemplateManagement/Factory/IConvertDataTemplateCollectionProviderFactory.cs

```diff
@@ -3,12 +3,15 @@
 // Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
 // -------------------------------------------------------------------------------------------------
 
+using Azure.Storage.Blobs;
 using Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders;
 
 namespace Microsoft.Health.Fhir.TemplateManagement.Factory
 {
-    public interface ITemplateProviderFactory
+    public interface IConvertDataTemplateCollectionProviderFactory
     {
-        public IConvertDataTemplateProvider GetTemplateProvider();
+        public IConvertDataTemplateCollectionProvider CreateTemplateCollectionProvider();
+
+        public IConvertDataTemplateCollectionProvider CreateTemplateCollectionProvider(BlobContainerClient blobContainerClient);
```

Comment:
for `CreateTemplateCollectionProvider(BlobContainerClient blobContainerClient);` it seems like this will require the caller to supply the blobContainerClient. I am not opposed to this but it seems like the caller will need to decide whether to call `CreateTemplateCollectionProvider();` or `CreateTemplateCollectionProvider(BlobContainerClient blobContainerClient);` which means some template collection provider logic and Options may need to reside in the caller rather than in the factory

Reply:
In a way, it is the same expectation as before, where the caller needed to decide whether to instantiate StorageTemplateCollectionProviderFactory() or DefaultTemplateCollectionProviderFactory() or future template collection collection providers. 
But yeah if we don't want that, we can have a generic TemplateCollectionConfiguration (maybe reuse the existing one) and depending on what options are set in there, the factory class can auto create the provider accordingly. But we would need to have some order of precedence in case the configuration had multiple provider info supplied (e.g. the config supplies both storage blob or github (future provider)). 
Thoughts?

Reply:
I do like not requiring a BlobContainerClient for the template collection provider if the intent to is to use this same provider when we add other options like Github in the future.

The ideal setup in my mind would be:
* An optional configuration class to control which providers are configured.
* A factory class.  The generic one would have the configuration as a constructor param injected via DI though we could also have versions of the factory that are specific to certain cases.  For example there could be an ACR version of the factory that is just for backwards compat with the existing FHIR service.
* For the actual IConvertDataTemplateCollectionProvider implementation, once we have a setup that allows multiple sources the logic for how they are resolved and the order can be internalized inside the implementation.  It could be controlled by options in the configuration that the factory will then pass along to appropriate collection provider that is being created.

All this is a long way of saying, I do believe we should try to avoid the explicit reference to the blob client in the factory interface signature since there could be situations where a blob source isn't valid (specifically when supporting backwards compatibility).  I know the old signature with out the blob client exists but as stated above I believe we can hide this from the create call with some tweaks.

Reply:
Yeah I did originally consider taking the container inputs as configuration options before, but since we were previously passing in the BlobContainerClient to the original StorageTemplateProviderFactory(), I left that as is as the expectation from the consumer. 
But agreed, I'll probably add a new 'StorageConfigurationOptions' class and have an optional property within TemplateCollectionConfiguration (which we already pass into this generic factory) and then internally create the appropriate providers depending on if this prop is set.

Personally, I don't like the idea of specific versions of factory, say each for ACR, Storage, Default, GitHub, etc. since I feel it could defeat the purpose of a factory class. Having a generic one that is responsible to create different template collection providers depending on provided inputs abstracts the creation logic without exposing to the client the different providers.

Also to clarify, this does not incorporate the previous ACR implementation. Prior to this change, it appeared that we have already created a segregation between the previous and new convert impl (e.g., creating new template collection provider and not extending the previous one), so going by that this new factory class doesn't deal with the prior providers.

Reply:
On the ACR implementation being separate, that is fair though we may want to reconsider that in the future it would be nice if the new interface could incorporate that scenario if necessary.

Multiple factories isn't needed and since right now the scenario is only for storage one implementation is fine.  If we prefer a model where factory contains all the possible sources and is driven by a config that is fine , though I do think there a case for dedicated scenario factories for each template source and there is common factory the invokes the correct ones based on the configuration.

Reply:
As discussed, incorporated the options to figure out which provider to return and internally creating the blob container client. 
@wi-y - as part of this I moved the relevant configuration classes from the convert repo here so the factory can just read these directly instead of the consumer parsing the config and creating the client. Let me know if there are any concerns.